doc/development_guide.md

```diff
@@ -12,134 +13,336 @@ a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow th
-1.  Add `<clone_root>` to your PYTHONPATH environment variable:
-
-    ##### Windows
-    ```
-    set PYTHONPATH=<clone_root>;%PYTHONPATH%
-    ```
-    ##### MacOS/Linux (bash)
-    ```
-    export PYTHONPATH=<clone_root>:${PYTHONPATH}
-    ```
-2.	Install the dependencies:
-    ```
-    python <clone_root>/dev_setup.py
-    ```
-## <a name="Configure_IDE"></a>3. Configuring your IDE
-
-#### Visual Studio Code (Any platform)
-
-1.	Install VS Code
+If not developing in a virtual environment, please proceed to [Install All Dependencies](#Install_All_Dependencies) 
+
+### 1. Create Virtual Environment
+Create a virtual environment in a subdirectory of your `<workspaceRoot>` (`<workspaceRoot>/env`, for example). Open command prompt or bash.
+
+##### Windows (Python3.x)
+```
+set PYLOC=<python3_location>
+set PATH=%PYLOC%\Scripts;%PYLOC%;%PATH%
+python -m venv <workspaceRoot>\env
+```
+##### Windows (Python2.x)
+```
+set PYLOC=<python2_location>
+set PATH=%PYLOC%\Scripts;%PYLOC%;%PATH%
+python -m pip install --upgrade pip
+python -m pip install --upgrade virtualenv
+python -m virtualenv <workspaceRoot>\env
+```
+##### MacOS (Python3.x)
+```
+python3 -m venv <workspaceRoot>/env
+```
+##### MacOS (Python2.x)
+```
+python -m venv <workspaceRoot>/env
+```
+##### Linux (Python3.x) -- Ubuntu for example
+```
+sudo apt-get update
+sudo apt-get install python3-pip
+sudo apt-get install python3-venv
+python3 -m venv <workspaceRoot>/env
+```
+##### Linux (Python2.x) -- Ubuntu for example
+```
+sudo apt-get update
+sudo apt-get install python-pip
+python -m pip install --upgrade pip
+python -m pip install virtualenv
+python -m virtualenv <workspaceRoot>/env
+```
+    
+### 2. Activate Virtual Environment
+##### Windows
+```
+<workspaceRoot>\env\scripts\activate.bat
+```
+##### MacOS/Linux
+```
+. <workspaceRoot>/env/bin/activate
+```
+
+### (To deactivate the Virtual Environment)
+##### Windows/MacOS/Linux
+```
+deactivate
+```
+
+## <a name="Build_mssql_cli"></a>3. Build mssql-cli
+General development steps that apply to both a virtual environment or a global environment.
+*If working in a virtual environment, do ensure the virtual environment is activated as the first command line of each following script.*
+
+##### Windows
+```
+<workspaceRoot>\env\scripts\activate.bat
+
+cd <workspaceRoot>
+python -m pip install --upgrade pip
+python build.py build
+
+set PYTHONPATH=<workspaceRoot>;%PYTHONPATH%
+python dev_setup.py
+```
+##### MacOS/Linux (Python3.x)
+```
+. <workspaceRoot>/env/bin/activate
+
+cd <workspaceRoot>
+python3 -m pip install --upgrade pip
+python3 build.py build
+
+export PYTHONPATH=<workspaceRoot>:${PYTHONPATH}
+python3 dev_setup.py
+```
+##### MacOS/Linux (Python2.x)
+```
+. <workspaceRoot>/env/bin/activate
+
+cd <workspaceRoot>
+python -m pip install --upgrade pip
+python build.py build
+
+export PYTHONPATH=<workspaceRoot>:${PYTHONPATH}
+python dev_setup.py

```

Comment:
Should we run dev_setup.py before we build?

Reply:
fixed.